### PR TITLE
Toggle option to enable/disable balance and Token rate checking for using third-party API

### LINKED ISSUE #16724: Make currency conversion features optional / toggleable
We hit coingecko and cryptocompare APIs for these features. We can make that optional.

**Current Mobile copy:**
Show balance and token price checker
We use Coingecko API to display your balance and token price. [Privacy Policy]

### PR DESCRIPTION
**Pull request by NiranjanaBinoy**
Fixes #16724 
## Explanation
We have added a new toggle to the `Privacy and Security` tab to control the use of third-party APIs like coinGecko and CryptoCompare for balance and Token rate checking.

### Before
Screenshot 2022-12-07 at 6 00 34 PM
Screenshot 2022-12-07 at 6 00 57 PM
Screenshot 2022-12-07 at 6 01 47 PM
Screenshot 2022-12-07 at 6 02 09 PM

### After
Screenshot 2022-12-07 at 5 59 02 PM
Screenshot 2022-12-07 at 5 58 28 PM
Screenshot 2022-12-07 at 5 59 31 PM
Screenshot 2022-12-07 at 5 59 52 PM

## Manual Testing Steps
Onboard v1:
  1. Load the application and import a wallet.
  2. The fiat values should not be visible - in eth-overview, token overview, or in asset/token lists
  3. Try to send a token, and fiat values for gas or total should not be visible.
  4. Now go to settings, search for `Show balance and token price checker` in the security tab and toggle it ON.
  5. Now, come back to the overview page and make sure Fiat values are visible.
  6. Again, Toggle OFF and make sure the Fiat values disappear from everywhere.
  
Onboarding V2:
1. Load the application and import a wallet.
2. Enable the `Show balance and token price checker` from the Privacy settings.
3. The app will display the fiat values.
4. reload the application and import the wallet again.
5. This time, without enabling `Show balance and token price checker`.
6. the Fiat values should not e displayed.

## Pre-merge author checklist

- [ ] I've clearly explained:
  - [ ] What problem this PR is solving
  - [ ] How this problem was solved
  - [ ] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

**Comment by danjm** [comment]
I think that with these modifications to this PR, we can greatly simplify the changes needed to controllers: https://github.com/MetaMask/metamask-extension/compare/optional-currency-conversion...optional-currency-conversion-alt?expand=1

I believe that with those changes, we will only need these changes to the controllers: https://github.com/MetaMask/controllers/compare/allow-disabled-config-tokenratescontroller?expand=1

**Comment by segun** [comment]
Hi @NiranjanaBinoy I was also working on this issue as part of my PR that fixes this broader Security and Privacy settings page #16727 and [this design](https://github.com/MetaMask/metamask-extension/issues/16727#issuecomment-1335121747) 

Can you confirm your PR is following this design as well

@danjm it seems there's some duplication of tasks here. For instance a PR already added this (merged to develop too recently).

```
Batch account balance requests
We batch accounts and query Infura to responsively show your balances. If you turn this off, only active accounts will be queried. Some dapps won't work unless you connect your wallet.
```

Which is part of the issue I'm working on as well.

Here's[ my PR](https://github.com/MetaMask/metamask-extension/pull/16756) that implemented the design.

As at the time I'm writing this, I have implemented everything according to that design except the last two. I hope to finish the designs today/tomorrow and then work on making sure the unit tests passed.

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -966,6 +1000,7 @@ export default class MetamaskController extends EventEmitter {
         // TODO: Handle failure to get conversion rate more gracefully
         console.error(error);
       }
+      // }
```
Comment (jpuri) [inline app/scripts/metamask-controller.js:1003]: 🔪

File: test/e2e/tests/edit-gas-fee.spec.js [hunk 2]
```diff
@@ -242,9 +242,6 @@ describe('Editing Confirm Transaction', function () {
         const editedTransactionAmount = editedTransactionAmounts[0];
         assert.equal(await editedTransactionAmount.getText(), '0.00021');
 
-        const editedTransactionFee = editedTransactionAmounts[1];
-        assert.equal(await editedTransactionFee.getText(), '0.00021');
-
```
[the remark this answers is not in the document]
Comment (NiranjanaBinoy) [reply test/e2e/tests/edit-gas-fee.spec.js:247]: it kept failing for me, and when I checked, there was some change in the CSS name because of which that value was not fetched.

[the remark this answers is not in the document]
Comment (danjm) [reply test/e2e/tests/edit-gas-fee.spec.js:247]: I don't think we should remove this check. We should figure out why it was failing.

File: test/e2e/tests/edit-gas-fee.spec.js [hunk 3]
```diff
@@ -150,7 +150,7 @@ describe('Editing Confirm Transaction', function () {
```
Comment (danjm) [inline test/e2e/tests/edit-gas-fee.spec.js:150]: @NiranjanaBinoy Why were these test changes needed?

File: test/e2e/tests/send-edit.spec.js [hunk 4]
```diff
@@ -145,7 +145,7 @@ describe('Editing Confirm Transaction', function () {
```
Comment (danjm) [inline test/e2e/tests/send-edit.spec.js:145]: @NiranjanaBinoy Why was this test change needed?

**Review by danjm — DISMISSED** [review]
Code changes look good, but I want to understand the e2e test changes before approving

File: app/scripts/controllers/detect-tokens.test.js [hunk 5]
```diff
@@ -65,15 +66,41 @@ describe('DetectTokensController', function () {
...
+                chainId: hexToDecimal(networkState.provider.chainId),
+              },
+            };
+            return cb(modifiedNetworkState);
+          }),
+      },
+      // {
+      //   provider: this.provider,
+      // },
+    );
```
[the remark this answers is not in the document]
Comment (jpuri) [reply app/scripts/controllers/detect-tokens.test.js:101]: Hey, this needs fix.

File: app/scripts/metamask-controller.js [hunk 6]
```diff
@@ -508,17 +522,37 @@ export default class MetamaskController extends EventEmitter {
...
+            { disabled: false },
+            false,
+            false,
+          );
+        } else if (!currUseCurrencyRateCheck && prevUseCurrencyRateCheck) {
+          this.currencyRateController.stop();
+          this.tokenRatesController.configure({ disabled: true }, false, false);
+        }
+      }, this.preferencesController.store.getState()),
+    );
```
Comment (jpuri) [inline app/scripts/metamask-controller.js:555]: Code above can better be written in `tokenRatesController`. I would avoid putting code in metamaskController.

**Comment by jpuri** [comment]
@NiranjanaBinoy : code changes look good, but I agree to @danjm feedback, plz ensure that changes in test cases are needed. Also will be nice to have new test cases added for the feature.

File: app/scripts/controllers/detect-tokens.test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (NiranjanaBinoy) [reply app/scripts/controllers/detect-tokens.test.js:101]: It was the first thing I removed, I might have sneaked back in somehow during rebase or something. will remove it now.

**Review by jpuri — DISMISSED** [review]

File: app/scripts/controllers/detect-tokens.js [hunk 7]
```diff
@@ -116,7 +116,7 @@ export default class DetectTokensController {
     const tokensToDetect = [];
     for (const tokenAddress in tokenListUsed) {
       if (
-        !this.tokenAddresses.find(({ address }) =>
+        !this.tokenAddresses.find((address) =>
```
Comment (adonesky1) [inline app/scripts/controllers/detect-tokens.js:119]: why was this change required?

File: ui/components/app/detected-token/detected-token-values/detected-token-values.js [hunk 8]
```diff
@@ -46,7 +46,7 @@ const DetectedTokenValues = ({
           {`${balanceString || '0'} ${token.symbol}`}
         </Typography>
         <Typography variant={TYPOGRAPHY.H7} color={COLORS.TEXT_ALTERNATIVE}>
-          {formattedFiatBalance || '$0'}
+          {formattedFiatBalance}
```
Comment (adonesky1) [inline ui/components/app/detected-token/detected-token-values/detected-token-values.js:49]: should we still show `$0` if the user has the `useCurrencyExchange` set to true but `formattedFiatBalance` is undefined?

File: app/_locales/en/messages.json [hunk 9]
```diff
@@ -678,6 +678,9 @@
   "close": {
     "message": "Close"
   },
+  "coingecko": {
+    "message": "Coingecko"
```
Comment (digiwand) [inline app/_locales/en/messages.json:682]:
```suggestion
    "message": "CoinGecko"
```

I think we'll want to update `Coingecko` -> `CoinGecko`
Screen Shot 2022-12-22 at 03 24 18

**Comment by adonesky1** [comment]
I'm not entirely sure what this `after` picture is meant to show:
206315310-d3b07342-b9e4-40cf-8911-dc6926671af3

**Review by digiwand — DISMISSED** [review]

**Review by digiwand — COMMENTED** [review]
--

**Comment by jpuri** [comment]
> I'm not entirely sure what this `after` picture is meant to show:

Hey @adonesky1 : this before picture shows this dollar value which is removed in after picture as user opted out for rate checking:

Screenshot 2022-12-22 at 10 46 30 AM

File: app/scripts/controllers/detect-tokens.js (hunk 7 above)
Comment (jpuri) [reply app/scripts/controllers/detect-tokens.js:119]: This change is correct, seems that previous code was wrong.

File: ui/components/app/detected-token/detected-token-values/detected-token-values.js (hunk 8 above)
Comment (jpuri) [reply ui/components/app/detected-token/detected-token-values/detected-token-values.js]: I have made change to still show default `$0` if `useCurrencyRateCheck` is true. Ideally is `useCurrencyRateCheck` is true and we do not get rate we show show something like `N/A` but we can check that in a separate PR.

File: app/scripts/controllers/detect-tokens.js [hunk 10]
```diff
@@ -116,7 +116,7 @@ export default class DetectTokensController {
...
           isEqualCaseInsensitive(address, tokenAddress),
         ) &&
         !this.hiddenTokens.find((address) =>
```
Comment (digiwand) [inline app/scripts/controllers/detect-tokens.js:122]:
```suggestion
        !this.hiddenTokens.find(({ address }) =>
```

File: app/scripts/controllers/detect-tokens.js (hunk 7 above)
Comment (digiwand) [reply app/scripts/controllers/detect-tokens.js:119]: lgtm

I think we'll want to update line 122 as well.  https://github.com/MetaMask/metamask-extension/pull/16772/files#r1055712836

**Comment by adonesky1** [comment]
> With toggle on and testing with the Goerli network, the "Send" token screen seems to show a duplicate value here that is not the checked token price

yeah if you don't have the show currency conversion on testnets turned on you get that extra row without the conversion but with that setting on it looks right: 

Screenshot 2022-12-22 at 11 55 35 AM

Looks like we shouldn't show it on a testnet unless both those settings are on.

**Comment by jpuri** [comment]
That extra row is caused but text wrapping when text width is large.

**Comment by adonesky1** [comment]
> That extra row is caused but text wrapping when text width is large.

Yep that's true. This is what it is on Polygon:
Screenshot 2022-12-22 at 11 58 30 AM

But we still shouldn't show the value twice @jpuri

File: app/scripts/controllers/detect-tokens.js (hunk 10 above)
Comment (NiranjanaBinoy) [reply app/scripts/controllers/detect-tokens.js:122]: `this.tokenAddresses` which is a list of token objects but `this.hiddenTokens` is a list of token addresses which is a list of strings.

**Review by jpuri — DISMISSED** [review]

File: ui/components/app/gas-details-item/gas-details-item.js [hunk 11]
```diff
@@ -39,14 +41,16 @@ const GasDetailsItem = ({ userAcknowledgedGasMissing = false }) => {
...
-            hideLabel={Boolean(useNativeCurrencyAsPrimaryCurrency)}
-          />
-        </div>
+        useCurrencyRateCheck && (
+          <div className="gas-details-item__currency-container">
+            <LoadingHeartBeat estimateUsed={estimateUsed} />
+            <UserPreferencedCurrencyDisplay
+              type={SECONDARY}
+              value={hexMinimumTransactionFee}
+              hideLabel={Boolean(useNativeCurrencyAsPrimaryCurrency)}
```
Comment (digiwand) [inline ui/components/app/gas-details-item/gas-details-item.js:50]: re: duplicate display value for testnet networks

\- starting a thread here to continue the convo above -

Using testnet networks and whether or not the "Primary Currency" setting is set to the current network currency or fiat, I am seeing the duplicate value in the `display-item` components.
Screen Shot 2023-01-06 at 7 47 37 PM

Screen Shot 2023-01-06 at 7 40 44 PM

Using Ethereum Mainnet, and whether or not the "Primary Currency" is set, I am seeing the fiat value displayed. This appears correct as mentioned in [above PR comments](https://github.com/MetaMask/metamask-extension/pull/16772#issuecomment-1363184217).
Screen Shot 2023-01-06 at 7 43 26 PM
Screen Shot 2023-01-06 at 7 43 52 PM


I'm thinking we can exclude the currency rate display for testnet currencies. Thoughts? @NiranjanaBinoy @adonesky1 @jpuri

Comment (digiwand) [reply ui/components/app/gas-details-item/gas-details-item.js:50]: Hey @NiranjanaBinoy, 

Sorry, I mixed up the relevant setting configs in my first comment. I think we can ignore `showFiatInTestnets`. Instead, the two relevant values are `showFiatConversionInTestnets` and `useCurrencyRateCheck`. I can reproduce the issue when `useCurrencyRateCheck` is true and `showFiatConversionInTestnets` is false. 

This recording has the reproduced example. It shows three states: 
| State in video | `useCurrencyRateCheck` | `showFiatConversionInTestnets` | Result |
| :---: | :---: | :---: | :---: |
| 1 | false | false |  ✅ |
| 2 | true | false |  duplicate value shown |
| 3 | true | true |  ✅ |


https://user-images.githubusercontent.com/20778143/211341448-6aeac8c2-c35a-43be-9718-b31ea7e25c02.mov

Are you able to reproduce it with these states?

Comment (digiwand) [reply ui/components/app/gas-details-item/gas-details-item.js:50]: I see and confirmed this is true. Thanks for clarifying! 
I created an issue ticket to remove the redundant value shown https://github.com/MetaMask/metamask-extension/issues/17118

File: ui/components/app/detected-token/detected-token-values/detected-token-values.js (hunk 8 above)
Comment (digiwand) [reply ui/components/app/detected-token/detected-token-values/detected-token-values.js]: It's not clear to me when `formattedFiatBalance` is falsy. Is there another case this may occur aside from not receiving a rate yet? If so, can we create a separate issue ticket and/or add a comment? If not, why not apply `N/A` instead of `$0` now? 

@jpuri @NiranjanaBinoy

**Review by digiwand — COMMENTED** [review]
looking good to me

There are 2 outstanding comments that are hidden away
https://github.com/MetaMask/metamask-extension/pull/16772#discussion_r1066263843
https://github.com/MetaMask/metamask-extension/pull/16772#discussion_r1050986139

File: app/scripts/metamask-controller.js (hunk 6 above)
Comment (NiranjanaBinoy) [reply app/scripts/metamask-controller.js:552]: Introducing these changes in `tokenRatecontroller` will need an introduction of `useCurrencyRateCheck` to `Preferencecontroller` in the core repo, which is redundant as we are not using a separate `Preferencecontroller` in the extension. This way, we can minimize the extension-specific code in the core repo.

File: ui/components/app/detected-token/detected-token-values/detected-token-values.js (hunk 8 above)
Comment (NiranjanaBinoy) [reply ui/components/app/detected-token/detected-token-values/detected-token-values.js]: I think for the `formattedFiatBalance` to be false the conversion rate has to be missing, or there is an issue with balance fetching while viewing the page. I don't remember exactly, but I think we might have added the `$0` after checking with the design time at the time of token detection v2. Are we sure changing that to `N/A` will not create any other confusion for the user?

Comment (NiranjanaBinoy) [reply ui/components/app/detected-token/detected-token-values/detected-token-values.js]: Will it be okay if we do that in a follow-up PR? This PR needs to go into v10.24 to fix an existing issue.

File: ui/pages/settings/security-tab/security-tab.component.js [hunk 12]
```diff
@@ -348,7 +353,7 @@ export default class SecurityTab extends PureComponent {
 
     return (
       <div
-        ref={this.settingsRefs[7]}
+        ref={this.settingsRefs[4]}
         className="settings-page__content-row"
         data-testid="advanced-setting-gas-fee-estimation"
```
Comment (digiwand) [inline ui/pages/settings/security-tab/security-tab.component.js:358]: [out of scope] I think we'll want to update this `data-testid` value

Comment (digiwand) [inline ui/pages/settings/security-tab/security-tab.component.js:358]: [out of scope] we'll want to update this `data-testid` value to reflect the related setting tab and setting

**Review by digiwand — DISMISSED** [review]

File: ui/components/app/detected-token/detected-token-values/detected-token-values.js (hunk 8 above)
Comment (kingdutch001) [reply ui/components/app/detected-token/detected-token-values/detected-token-values.js]: Warnings
⚠️	
Libraries/Components/Button.js#L14 - Libraries/Components/Button.js line 14 – Requires should be sorted alphabetically (lint/sort-imports)

**Review by digiwand — DISMISSED** [review]

File: app/scripts/controllers/detect-tokens.js (hunk 10 above)
Comment (danjm) [reply app/scripts/controllers/detect-tokens.js:122]: @NiranjanaBinoy I just want to confirm. Here you say "this.tokenAddresses which is a list of token objects", but above you changed the code to `!this.tokenAddresses.find((address) =>`. This seems to be a mismatch. Was there a mistake somewhere (either in the code or in your comment)?

**Review by digiwand — DISMISSED** [review]
🙌🏼

**Review by jpuri — DISMISSED** [review]

File: app/scripts/metamask-controller.js [hunk 13]
```diff
@@ -505,17 +519,37 @@ export default class MetamaskController extends EventEmitter {
...
                 chainId: hexToDecimal(networkState.provider.chainId),
               },
             };
             return cb(modifiedNetworkState);
           }),
       },
-      undefined,
+      {
+        disabled:
+          !this.preferencesController.store.getState().useCurrencyRateCheck,
```
Comment (danjm) [inline app/scripts/metamask-controller.js:532]: we might want to explicitly check that this is `false`, so that this is not disabled for people who have never touched the toggle

File: app/scripts/controllers/preferences.js [hunk 14]
```diff
@@ -42,6 +42,7 @@ export default class PreferencesController {
       // set to false will be using the static list from contract-metadata
       useTokenDetection: false,
       useNftDetection: false,
+      useCurrencyRateCheck: false,
```
Comment (danjm) [inline app/scripts/controllers/preferences.js:45]: I think we want to default this to `true`. Current behaviour in prod today is the same as if this was true

**Review by danjm — CHANGES_REQUESTED** [review]
overall looks good. Only one significant comment from me

File: app/scripts/controllers/detect-tokens.js (hunk 10 above)
Comment (NiranjanaBinoy) [reply app/scripts/controllers/detect-tokens.js:122]: The mistake was in my comment; in the state, we are saving `tokens` as a list of objects, but in the constructor of `DetectTokensController,` we are filtering out the address to the `this.tokenAddresses` list. https://github.com/MetaMask/metamask-extension/blob/e73afb8451e7b055091f200ce00826ade6d0b924/app/scripts/controllers/detect-tokens.js#L74

**Review by danjm — DISMISSED** [review]
LGTM!

**Review by jpuri — DISMISSED** [review]

**Review by digiwand — DISMISSED** [review]

**Comment by seaona** [comment]
When I do the Onboarding V1, I can see that the fiat values are there and the toggle is On by default. Should this be the case?

https://user-images.githubusercontent.com/54408225/212341339-a8b87523-2cf3-4265-80e5-791a5bc94d48.mp4

**Review by seaona — CHANGES_REQUESTED** [review]
When I do the Onboarding V1, I can see that the fiat values are there and the toggle is On by default. Should this be the case?

https://user-images.githubusercontent.com/54408225/212341509-88f01942-f3b7-4756-8779-84327e6c5f25.mp4

**Comment by NiranjanaBinoy** [comment]
> When I do the Onboarding V1, I can see that the fiat values are there and the toggle is On by default. Should this be the case?

@seaona  Yes, we have updated the toggle to be ON by default in [this commit](https://github.com/MetaMask/metamask-extension/pull/16772/commits/cc1948d55045a7a692c3531b5c5c8db2ed064637), this was a last-minute change and I forgot to change the PR description after that

**Comment by NiranjanaBinoy** [comment]
@seaona Nice catch; I have updated the changes for swaps now; kindly verify it.

File: ui/pages/swaps/view-quote/view-quote.js [hunk 15]
```diff
@@ -1022,10 +1024,12 @@ export default function ViewQuote() {
...
                 maxFee: maxFeeInEth,
               }}
-              secondaryFee={{
-                fee: feeInFiat,
-                maxFee: maxFeeInFiat,
-              }}
+              secondaryFee={
+                useCurrencyRateCheck && {
+                  fee: feeInFiat,
+                  maxFee: maxFeeInFiat,
```
Comment (digiwand) [inline ui/pages/swaps/view-quote/view-quote.js:1030]: should we keep the maxFee here and show it in the current currency? 

w/ currency rate setting turned off:
Screen Shot 2023-01-16 at 11 18 17 PM

w/ currency rate setting turned on: 
Screen Shot 2023-01-16 at 11 17 11 PM

Comment (NiranjanaBinoy) [reply ui/pages/swaps/view-quote/view-quote.js:1030]: When the currency rate checking is disabled, we will stop fetching the conversion rates, so the max fee displayed above will not be accurate.

Comment (danjm) [reply ui/pages/swaps/view-quote/view-quote.js:1030]: yeah, best not to show fiat values when the currency rating check is disabled

**Review by danjm — DISMISSED** [review]
Looks good to me

File: ui/pages/swaps/view-quote/view-quote.js (hunk 15 above)
Comment (digiwand) [reply ui/pages/swaps/view-quote/view-quote.js:1030]: still wrapping my head around the logic. I think I understand that turning off the currency rate setting will stop the polling process. I thought this was related to fetching currency rates to check against the network currency, and that this is separate from fetching the current network currency data. `getRenderableNetworkFeesForQuote` in the screenshot below appears to provide the maxFee in Fiat and the currency network currency. Is the `maxFeeInEth` impacted by the new setting? 

Screen Shot 2023-01-17 at 1 14 05 AM
Screen Shot 2023-01-17 at 1 20 50 AM

Comment (digiwand) [reply ui/pages/swaps/view-quote/view-quote.js:1030]: @NiranjanaBinoy explained to me that turning off the setting
> blocks the polling for conversion rates from CurrencyRateController and TokensRateController. So if the toggle is OFF the update values for conversion rates will not be pulled in and any value that is still in state will be out dated.

I think I see how the values shown can be outdated now.

Thanks for the explanation @NiranjanaBinoy! marking this as resolved

Comment (digiwand) [reply ui/pages/swaps/view-quote/view-quote.js:1030]: @NiranjanaBinoy explained to me that turning off the setting

blocks the polling for conversion rates from CurrencyRateController and TokensRateController. So if the toggle is OFF the update values for conversion rates will not be pulled in and any value that is still in state will be outdated.

I think I see how the values shown can be outdated now. Thanks for this explanation @NiranjanaBinoy! 

I'm going to take a bit more time to understand this more.

Comment (NiranjanaBinoy) [reply ui/pages/swaps/view-quote/view-quote.js]: @digiwand I have updated the code to use the maxFeeInEth for max Fee in the swaps page. After discussing with @danjm, it was decided to add the maxFeeInEth when the toggle is OFF. Thanks for raising the concern ❤️

**Review by pedronfigueiredo — APPROVED** [review]

File: ui/pages/swaps/view-quote/view-quote.js (hunk 15 above)
Comment (digiwand) [reply ui/pages/swaps/view-quote/view-quote.js]: awesome! Thanks for looking into it more w/ Dan. LGTM ❤️

**Review by digiwand — APPROVED** [review]

**Review by danjm — APPROVED** [review]

**Review by seaona — APPROVED** [review]
